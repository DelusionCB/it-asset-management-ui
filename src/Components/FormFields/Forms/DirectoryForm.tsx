import React, {useEffect, useState} from 'react';
import {Row} from 'reactstrap';
import {
    directoryDefaultValues,
    extendedDirectoryDataProps,
} from '../../../Types/types.directories';
import TextField from '../../CustomComponents/TextField';
import CSelect from '../../CustomComponents/CustomSelect';
import {handleDirectoryChange, handleDirectorySelect, debouncedCheck, handleClear} from '../../../Utils/helpers';
import {useTranslation} from 'react-i18next';
import ActionButtons from '../../ActionButtons/ActionButtons';
import VisibilitySelector from '../../CustomComponents/VisibilitySelector';

interface FieldsStates {
    isDisabled: (is: boolean) => void
    type: string
}

function DirectoryForm ({isDisabled, type}: FieldsStates): JSX.Element {
    const [values, setValues] = useState<extendedDirectoryDataProps>(directoryDefaultValues)
    const {t} = useTranslation()

    function handleChange (e: React.ChangeEvent<HTMLInputElement>): void {
        handleDirectoryChange(e, values, setValues);
    }

    function handleSelect (value: {} | [] | undefined | null, id: string): void {
        handleDirectorySelect(value, id, values, setValues);
    }

    useEffect(() => {
        // Invoke debouncedCheck with updated values and directoryDefaultValues
        debouncedCheck(values, directoryDefaultValues, isDisabled);

        // Cleanup function to cancel the debounce on unmount
        return () => {
            debouncedCheck.cancel();
        };
    }, [values]);

    useEffect(() => {
        if (values.visibility === 'hidden' || values.visibility === 'draft') {
            handleClear<extendedDirectoryDataProps>(
                values,
                [],
                ['applications', 'servers', 'services'],
                setValues,
            )
        }
    }, [values.visibility])

    return (
        <>
            <Row>
                <h1>{t('directory.create')}</h1>
            </Row>
            <Row>
                <h2>{t('directory.visibility')}</h2>
            </Row>
            <Row>
                <VisibilitySelector
                    value={values.visibility}
                    onChange={(e) => { handleChange(e); }}
                    disabled={false}
                    id='visibility'
                    label={'values.visibility'}
                    invalid={false}
                    options={['draft', 'hidden', 'published']}
                    placeholder={'placeholder.visibility'}
                />
            </Row>
            <Row>
                <h2>{t('directory.description')}</h2>
            </Row>
            <Row>
                <TextField
                    id='name'
                    disabled={false}
                    invalid={false}
                    label={'values.name'}
                    onChange={(e) => { handleChange(e); }}
                    placeholder={'placeholder.name'}
                    value={values.name}
                    type='text'
                />
            </Row>
            <Row>
                <TextField
                    id={'description'}
                    disabled={false}
                    invalid={false}
                    type='textarea'
                    label={'values.description'}
                    onChange={(e) => { handleChange(e) }}
                    placeholder={'placeholder.description'}
                    value={values.description}
                />
            </Row>
            <Row>
                <h2>{t('directory.dependency')}</h2>
            </Row>
            <Row>
                <CSelect
                    label={'values.applications'}
                    endpoint='application'
                    id={'applications'}
                    placeholder={'placeholder.applications'}
                    disabled={values.visibility === 'hidden' || values.visibility === 'draft'}
                    invalid={false}
                    isMulti={true}
                    value={values.applications}
                    onChange={(value, id) => { handleSelect(value, id); }}
                />
            </Row>
            <Row>
                <CSelect
                    label={'values.servers'}
                    endpoint='server'
                    id={'servers'}
                    placeholder={'placeholder.servers'}
                    disabled={values.visibility === 'hidden' || values.visibility === 'draft'}
                    invalid={false}
                    isMulti={true}
                    value={values.servers}
                    onChange={(value, id) => { handleSelect(value, id); }}
                />
            </Row>
            <Row>
                <CSelect
                    label={'values.services'}
                    endpoint='service'
                    id={'services'}
                    placeholder={'placeholder.services'}
                    disabled={values.visibility === 'hidden' || values.visibility === 'draft'}
                    invalid={false}
                    isMulti={true}
                    value={values.services}
                    onChange={(value, id) => { handleSelect(value, id); }}
                />
            </Row>
            <ActionButtons
                values={values}
                actions={['create']}
                type={type}
            />
        </>
    );
}

export default DirectoryForm;
