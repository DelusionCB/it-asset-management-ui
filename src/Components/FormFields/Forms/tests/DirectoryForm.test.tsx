import React from 'react';
import * as TestRenderer from 'react-test-renderer';
import DirectoryForm from '../DirectoryForm'
import {BrowserRouter} from 'react-router-dom'
import '@testing-library/jest-dom'

describe('<DirectoryFields />', () => {
    let renderer: TestRenderer.ReactTestRenderer;

    // Render the component before tests
    beforeAll(() => {
        renderer = TestRenderer.create(
            <BrowserRouter>
                <DirectoryForm type='directory' isDisabled={jest.fn()} />
            </BrowserRouter>,
        );
    });
    describe('Snapshot', () => {
        it('renders correctly', () => {
            expect(renderer.toJSON()).toMatchSnapshot();
        })
    })
})
