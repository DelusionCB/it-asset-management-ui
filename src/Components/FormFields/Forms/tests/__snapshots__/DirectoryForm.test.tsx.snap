// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DirectoryFields /> Snapshot renders correctly 1`] = `
[
  <div
    className="row"
  >
    <h1>
      directory.create
    </h1>
  </div>,
  <div
    className="row"
  >
    <h2>
      directory.visibility
    </h2>
  </div>,
  <div
    className="row"
  >
    <div
      className="textfield col"
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="visibility"
        >
          values.visibility
        </label>
        <select
          aria-invalid={false}
          className="form-select"
          disabled={false}
          id="visibility"
          name="visibility"
          onChange={[Function]}
          placeholder="placeholder.visibility"
          value="published"
        >
          <option
            value="draft"
          >
            visibility.draft
          </option>
          <option
            value="hidden"
          >
            visibility.hidden
          </option>
          <option
            value="published"
          >
            visibility.published
          </option>
        </select>
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <h2>
      directory.description
    </h2>
  </div>,
  <div
    className="row"
  >
    <div
      className="textfield col"
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="name"
        >
          directory.values.name
        </label>
        <input
          aria-invalid={false}
          className="form-control"
          disabled={false}
          id="name"
          name="name"
          onChange={[Function]}
          placeholder="placeholder.name"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="textfield col"
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="description"
        >
          directory.values.description
        </label>
        <textarea
          aria-invalid={false}
          className="form-control"
          disabled={false}
          id="description"
          name="description"
          onChange={[Function]}
          placeholder="placeholder.description"
          value=""
        />
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <h2>
      directory.dependency
    </h2>
  </div>,
  <div
    className="row"
  >
    <div>
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="applications"
        >
          directory.values.applications
        </label>
        <div
          className=" css-b62m3t-container"
          onKeyDown={[Function]}
        >
          <span
            className="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            className="css-1f43avz-a11yText-A11yText"
          />
          <div
            className="option-search__control css-13cymwt-control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="option-search__value-container option-search__value-container--is-multi css-1fdsijx-ValueContainer"
            >
              <div
                className="option-search__placeholder css-1jqq78o-placeholder"
                id="react-select-2-placeholder"
              >
                placeholder.applications
              </div>
              <div
                className="option-search__input-container css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-2-placeholder"
                  aria-expanded={false}
                  aria-haspopup={true}
                  aria-invalid={false}
                  aria-label="placeholder.applications"
                  aria-labelledby="applications"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  className="option-search__input"
                  disabled={false}
                  id="applications"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  role="combobox"
                  spellCheck="false"
                  style={
                    {
                      "background": 0,
                      "border": 0,
                      "color": "inherit",
                      "font": "inherit",
                      "gridArea": "1 / 2",
                      "label": "input",
                      "margin": 0,
                      "minWidth": "2px",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "100%",
                    }
                  }
                  tabIndex={0}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="option-search__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                className="option-search__indicator-separator css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                className="option-search__indicator option-search__dropdown-indicator css-1xc3v61-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div>
            <input
              name="applications"
              type="hidden"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <div>
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="servers"
        >
          directory.values.servers
        </label>
        <div
          className=" css-b62m3t-container"
          onKeyDown={[Function]}
        >
          <span
            className="css-1f43avz-a11yText-A11yText"
            id="react-select-3-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            className="css-1f43avz-a11yText-A11yText"
          />
          <div
            className="option-search__control css-13cymwt-control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="option-search__value-container option-search__value-container--is-multi css-1fdsijx-ValueContainer"
            >
              <div
                className="option-search__placeholder css-1jqq78o-placeholder"
                id="react-select-3-placeholder"
              >
                placeholder.servers
              </div>
              <div
                className="option-search__input-container css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-3-placeholder"
                  aria-expanded={false}
                  aria-haspopup={true}
                  aria-invalid={false}
                  aria-label="placeholder.servers"
                  aria-labelledby="servers"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  className="option-search__input"
                  disabled={false}
                  id="servers"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  role="combobox"
                  spellCheck="false"
                  style={
                    {
                      "background": 0,
                      "border": 0,
                      "color": "inherit",
                      "font": "inherit",
                      "gridArea": "1 / 2",
                      "label": "input",
                      "margin": 0,
                      "minWidth": "2px",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "100%",
                    }
                  }
                  tabIndex={0}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="option-search__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                className="option-search__indicator-separator css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                className="option-search__indicator option-search__dropdown-indicator css-1xc3v61-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div>
            <input
              name="servers"
              type="hidden"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <div>
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="services"
        >
          directory.values.services
        </label>
        <div
          className=" css-b62m3t-container"
          onKeyDown={[Function]}
        >
          <span
            className="css-1f43avz-a11yText-A11yText"
            id="react-select-4-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            className="css-1f43avz-a11yText-A11yText"
          />
          <div
            className="option-search__control css-13cymwt-control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="option-search__value-container option-search__value-container--is-multi css-1fdsijx-ValueContainer"
            >
              <div
                className="option-search__placeholder css-1jqq78o-placeholder"
                id="react-select-4-placeholder"
              >
                placeholder.services
              </div>
              <div
                className="option-search__input-container css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-4-placeholder"
                  aria-expanded={false}
                  aria-haspopup={true}
                  aria-invalid={false}
                  aria-label="placeholder.services"
                  aria-labelledby="services"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  className="option-search__input"
                  disabled={false}
                  id="services"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  role="combobox"
                  spellCheck="false"
                  style={
                    {
                      "background": 0,
                      "border": 0,
                      "color": "inherit",
                      "font": "inherit",
                      "gridArea": "1 / 2",
                      "label": "input",
                      "margin": 0,
                      "minWidth": "2px",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "100%",
                    }
                  }
                  tabIndex={0}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="option-search__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                className="option-search__indicator-separator css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                className="option-search__indicator option-search__dropdown-indicator css-1xc3v61-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div>
            <input
              name="services"
              type="hidden"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="action-buttons"
  >
    <button
      className="action-button btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      button.cancel
    </button>
    <button
      className="action-button btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      button.create
    </button>
  </div>,
]
`;
